### DISCUSSION TITLE
Error: Unable to find component

### DISCUSSION THREAD

**Discussion by agentphoenix** [awcodes/scribble#11] [General]
I've been trying to get Scribble up and running as a replacement for TipTapEditor. Most of my conversion has gone smoothly on the first custom block, but I'm running into an issue where an error gets thrown when trying to select the block I want to add. It's telling me that it's unable to find the component with the name `hero-split-with-image`.

I've tried manually registering the component (which I do in my app because I'm not using a traditional app structure), but that doesn't fix the issue. Any ideas what might causing this issue? I didn't need to register anything with TipTapEditor, but maybe something has changed which requires a change.

```php
// AppServiceProvider.php

public function boot(): void
{
  ScribbleFacade::registerTools($this->app[BlockManager::class]->blocks());
}

protected function registerBlocks(): void
{
  $blockManager = new BlockManager();

  $blockManager->registerPageBlocks([
    SplitWithImageHeroBlock::make(),
  ]);

  $this->app->scoped(BlockManager::class, fn () => $blockManager);
}
```

```php
// My profile for this particular instance

class PageBuilder extends ScribbleProfile
{
  public static function bubbleTools(): array
  {
    return [];
  }

  public static function suggestionTools(): array
  {
    return ScribbleFacade::getTools([
      'hero-split-with-image',
    ])->toArray();
    
    // This is what I'd like to do, but having some issues with it right now...
    return ScribbleFacade::getTools(
      app(BlockManager::class)->pageBlocks()
    )->toArray();
  }

  public static function toolbarTools(): array
  {
    return [];
  }
}
```

```php
// My custom block

abstract class HeroBlock extends ScribbleTool
{
  protected function baseConfiguration(): self
  {
    return $this
      ->icon('tabler-star')
      ->type(ToolType::Block);
  }
}

class SplitWithImageHeroBlock extends HeroBlock
{
  protected function setUp(): void
  {
    $this->baseConfiguration()
      ->label('Hero - Split w/ image')
      ->identifier('hero-split-with-image')
      ->optionsModal(SplitWithImageHeroBlockOptions::class)
      ->renderedView('pages.pages.blocks.hero.split-with-block-image')
      ->editorView('pages.pages.blocks.hero.split-with-block-image-preview');
  }
}

class SplitWithImageHeroBlockOptions extends ScribbleModal
{
  public ?string $header = 'Hero - Split w/ Image Options';

  public ?string $identifier = 'hero-split-with-image';

  public static function getSlideDirection(): SlideDirection
  {
    return SlideDirection::Left;
  }

  public static function getMaxWidth(): MaxWidth
  {
    return MaxWidth::ExtraLarge;
  }

  public function mount(): void
  {
    //
  }

  public function getFormFields(): array
  {
    return [];
  }
}
```

```php
// How I'm using it in my form component

ScribbleEditor::make('blocks')->profile(PageBuilder::class),
```

**Comment by awcodes** [comment]
If you dd(ScribbleFacade::getTools()->toArray()) what does it show?

**Comment by awcodes** [comment]
Might be an issue with the facade too. Try it with the ScribbleManager class directly.

**Reply by agentphoenix** [reply]
Same issue with the `ScribbleManager`.

<img width="686" alt="Screenshot 2024-03-05 at 8 10 23 PM" src="https://github.com/awcodes/scribble/assets/14422/923ba5b1-0fc9-403b-b50b-836c8043ca57">

Since tools are being registered as full objects, what would be the best way to handle this? Do I need to spin through my custom blocks and return an array of identifiers?

**Reply by agentphoenix** [reply]
`ScribbleManager` line 57 is the culprit. The call to `only` under the hood uses `array_flip`. Not saying it needs to be changed, just was doing a little debugging.

**Reply by awcodes** [reply]
Can you give me a stack trace or flair share url?

**Reply by agentphoenix** [reply]
https://flareapp.io/share/87nd2lGm

**Comment by awcodes** [comment]
Boot() might be too late to register the tools too.

**Reply by awcodes** [reply]
Yea, booted is going to be too late to register the tools since the package is autoregistering the livewire components in its package booted method. So tool registration needs to be done in the register() method of the service provider.

Probably need to update the docs.

**Reply by agentphoenix** [reply]
Moving the tool registration to `register` did the trick. Thanks!

**Comment by awcodes** [comment]
any way you could share your `BlockManager` class?

**Reply by agentphoenix** [reply]
```php
class BlockManager
{
    protected array $formBlocks = [];

    protected array $pageBlocks = [];

    public function blocks(): array
    {
        return array_merge(
            $this->pageBlocks,
            $this->formBlocks
        );
    }

    public function formBlocks(): array
    {
        return $this->formBlocks;
    }

    public function pageBlocks(): array
    {
        return $this->pageBlocks;
    }

    public function registerFormBlock(string $blockClass): self
    {
        $this->formBlocks[] = $blockClass;

        return $this;
    }

    public function registerFormBlocks(array $blocks): self
    {
        $this->formBlocks = array_merge($this->formBlocks, $blocks);

        return $this;
    }

    public function registerPageBlock(string $blockClass): self
    {
        $this->pageBlocks[] = $blockClass;

        return $this;
    }

    public function registerPageBlocks(array $blocks): self
    {
        $this->pageBlocks = array_merge($this->pageBlocks, $blocks);

        return $this;
    }
}
```

**Comment by awcodes** [comment]
Ok, I think this isn't working because your block manager isn't returning an array of string identifiers.

**Reply by agentphoenix** [reply]
Yeah I figured that was the case. I re-wrote the `pageBlocks` method to return an array of string identifier and it's working now. Thanks!
